### ISSUE TITLE
Messages Acked from SubscribeSync cause Drain() to timeout

### ISSUE THREAD

**Issue by vdwees** [nats-io/nats.go#790]
#### Versions of `nats.go` and the `nats-server` if one was involved:
nats.go v1.11.1-0.20210623165838-4b75fc59ae30, freshly started nats nats:2.3.4-alpine docker image

#### OS/Container environment:
pop os

#### Steps or code to reproduce the issue:
```go
func main() {
	drainedc := make(chan bool)
	nc, _ := nats.Connect("localhost:4222", // fresh nats nats:2.3.4-alpine docker image
		nats.DrainTimeout(time.Second*20),
		nats.ClosedHandler(func(_ *nats.Conn) {
			drainedc <- true
		}),
	)
	js, _ := nc.JetStream()
	streamConfig := nats.StreamConfig{
		Name:      "ORDERS",
		Subjects:  []string{"ORDERS.*"},
	}
	js.AddStream(&streamConfig)

	n, _ := strconv.Atoi(os.Args[1])
	for i := 0; i < n; i++ {
		js.PublishAsync("ORDERS.scratch", []byte("hello"))
	}
	<-js.PublishAsyncComplete()

	sub, _ := js.SubscribeSync("ORDERS.*", nats.ManualAck())
	var messages []*nats.Msg
	for i := 0; i < n; i++ {
		msg, _ := sub.NextMsg(time.Second)
		messages = append(messages, msg)
	}
	fmt.Printf("%d\n", len(messages))
	for _, m := range messages {
		m.Ack()
	}
	nc.Drain()
	<-drainedc
}
```

#### Expected result:

Run a few times:
```
> ./test 10
10
> ./test 10
10
> ./test 3
3
> ./test 1
1
```

#### Actual result:

Run a few times:
```
> ./test 10
10
> ./test 10
10
< wait 20s for nats drain to time out >
nats: draining connection timed out on connection [23]
> ./test 3
3
< wait 20s for nats drain to time out >
nats: draining connection timed out on connection [11]
> ./test 1
1
< wait 20s for nats drain to time out >
nats: draining connection timed out on connection [14]
```

**Comment by kozlovic** [comment]
If you delete the stream at the beginning of the test, I can't reproduce the issue. Without the delete, the test keep adding messages to the stream and when starting the subscription, the server may send to the sub more than you are going to call NextMsg() on. Since you use a sync subscription and you don't drain the subscription, I think what you observe is expected, but I may be wrong. @derekcollison Is my understanding correct?

**Comment by derekcollison** [comment]
Yes so on subsequent runs the stream already has the messages from the previous run and since you only process the messages via the consumer for the current run there are messages left over that need to be processed but you are not calling NextMsg anymore so they can't be drained per se.

You could create the stream each time as Ivan suggested. You could also Purge() the stream after you add it, or you could make the stream InterestPolicy so as the messages are consumed they will be deleted from the stream. Any of those would work.

**Comment by vdwees** [comment]
Thanks, this is actually super helpful in getting to the bottom of this!

Yes, in my code I have a WorkQueuePolicy, and if it never restarts its fine, but its running in k8s so it should be hardened against interruptions like being killed (with a short grace period) and restarted immediately on a different node while many messages are are in the backlog, for example.

Would the way forward be to use the js.Subcribe approach with the callback sending the message immediately to a large (larger than MaxAckPending()) buffered channel? I have tested this, when using `AckSync` the process can shutdown instantly, although its noticeably slower than using `Ack`. If I spawn more than one consumer (as my production code does) and use `Ack` instead of `AckSync` I can still hit the timeout message, e.g:

Running this with n=100_000,
```go
func main() {
	drainedc := make(chan bool)
	nc, _ := nats.Connect("localhost:4222", // fresh nats nats:2.3.4-alpine docker image
		nats.DrainTimeout(time.Second*20),
		nats.ClosedHandler(func(_ *nats.Conn) {
			drainedc <- true
		}),
	)
	js, _ := nc.JetStream()
	streamConfig := nats.StreamConfig{
		Name:      "ORDERS",
		Subjects:  []string{"ORDERS.*"},
		Retention: nats.WorkQueuePolicy,
	}
	js.AddStream(&streamConfig)

	n, _ := strconv.Atoi(os.Args[1])
	for i := 0; i < n; i++ {
		js.PublishAsync("ORDERS.scratch0", []byte(fmt.Sprintf("l=%d", i)))
		js.PublishAsync("ORDERS.scratch1", []byte(fmt.Sprintf("m=%d", i)))
		js.PublishAsync("ORDERS.scratch2", []byte(fmt.Sprintf("n=%d", i)))
	}
	<-js.PublishAsyncComplete()

	messages := make(chan *nats.Msg, 65535*3)

	js.Subscribe(
		"ORDERS.scratch0",
		func(m *nats.Msg) { messages <- m },
		nats.ManualAck(),
		nats.AckWait(time.Minute*5),
	)
	js.Subscribe(
		"ORDERS.scratch1",
		func(m *nats.Msg) { messages <- m },
		nats.ManualAck(),
		nats.AckWait(time.Minute*5),
	)
	js.Subscribe(
		"ORDERS.scratch2",
		func(m *nats.Msg) { messages <- m },
		nats.ManualAck(),
		nats.AckWait(time.Minute*5),
	)

	cancelc := make(chan bool)
	time.AfterFunc(time.Second*5, func() { cancelc <- true })
	ticker := time.NewTicker(time.Second)

	for {
		select {
		case <-ticker.C:
			fmt.Printf("len/cap messagec: %d/%d\n", len(messages), cap(messages))
		case <-drainedc:
			return
		case <-cancelc:
			nc.Drain()
		case m := <-messages:
			fmt.Println(string(m.Data))
			m.Ack()
		}
	}
}
```
prints:

```
...
n=99991
n=99992
n=99993
n=99994
n=99995
n=99996
n=99997
n=99998
n=99999
len/cap messagec: 0/196605
len/cap messagec: 0/196605
context deadline exceeded on connection [4]
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
context deadline exceeded on connection [4]
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
context deadline exceeded on connection [4]
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
len/cap messagec: 0/196605
```

**Comment by vdwees** [comment]
I am not seeing the errors you mention, and the Ack() calls don't return errors. In the logs I see that the last messages were Acked before the Drain() is called. I think I have to agree with you that its likely that there simply not enough time to process this volume of async Acks.

I mentioned that AckSync() is actually fine, but it is substantially slower than Ack(). Is there anything in the nats api similar to PublishAsyncComplete() that can be used to ensure that the async Acks are flushed? Or maybe a counter of pending acks? If not, I think it would be helpful to have something like this so that it is possible for the nats server to apply some back-pressure on the ack backlog. I suspect with normal nats this might never be a problem, but with Jetstream (and perhaps with work queues in particular) it seems like it is quite easy to write a service in go that can have a unbounded growing ack backlog.

**Comment by derekcollison** [comment]
The acks are scoped to the consumer which is a different one since its not durable. Those acks also have no affect on the stream unless running in interest based retention policy.

More over, since the consumer will process and ack only N messages when the stream has more, the second run will already cause issues with drain since we need the messages to be processed to drain effectively and you are no longer processing any messages after the initial N, even though on the second run you have 2N messages in the stream.

**Comment by vdwees** [comment]
Thanks for your feedback! The changes I made were:

1. Switch from `SubscribeSync` to `Subscribe` and call `Drain()` on the subscribers immediately on receiving the SIGINT. This allows the unprocessed message queue to empty so that there are no unprocessed messages when the final `nc.Drain()` is called. As @derekcollison pointed out, unprocessed messages also contributed to the timeouts.
1. Increase the timeout to 20 seconds so that my many thousands of async `Ack()`s can drain within the `DrainTimeout`. @kozlovic was right that `nc.Drain()` was simply running out of time. I also tested with different stores and found that in the case where there are simply too many Acks to fit in 20 seconds, I can use `MemoryStorage` instead of `FileStorage` and it can handle more than an order of magnitude more Acks.

I think its ok to close this issue.

@derekcollison I was introduced to NATS when you showed up on the gotime podcast. Thanks!
